### ISSUE TITLE
Consider setting filetype ‘clojure’ for file extension ‘edn’

### ISSUE THREAD

**Issue by glts** [vim/vim#863]
[edn](https://github.com/edn-format/edn) syntax is a subset of
[Clojure](http://clojure.org/) syntax.

Files in the edn format commonly have the file extension `.edn`.

Currently Vim assigns file type `edif` to `.edn` files. Not sure how
much this is used, so please consider reassigning this file extension to
the `clojure` file type.

**Comment by brammool** [comment]
> [EDN](https://github.com/edn-format/edn) syntax is a subset of
> [Clojure](http://clojure.org/) syntax.
> 
> Files in the EDN format commonly have the file extension `.edn`.
> 
> Currently Vim assigns file type `edif` to `.edn` files. Not sure how
> much this is used, so please consider reassigning this file extension to
> the `clojure` file type.

It does not seem commonly used: http://fileinfo.com/extension/edn

If it's possible to decide about the type from the contents, we can
write a function to do that.

## 

From "know your smileys":
 :q vi user saying, "How do I get out of this damn emacs editor?"

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\
\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by glts** [comment]
Looking at GitHub, edn seems to be rather more common than EDIF.
Wikipedia says that the last activity on EDIF was in the 1990s. edn
usage on the other hand is thriving in the Clojure community. My guess
is that nowadays edn is a better default choice for `.edn` files.

Anyway, in the Wikipedia article it looks like EDIF uses only round
brackets. edn uses square and curly brackets as well, so perhaps that
could be used to distinguish …

**Comment by glts** [comment]
The following patch adds support for the edn file format.

```patch
diff --git a/runtime/filetype.vim b/runtime/filetype.vim
index 9c9c808..dba6784 100644
--- a/runtime/filetype.vim
+++ b/runtime/filetype.vim
@@ -676,8 +676,14 @@ au BufNewFile,BufRead *.dtd			setf dtd
 " DTS/DSTI (device tree files)
 au BufNewFile,BufRead *.dts,*.dtsi		setf dts
 
-" EDIF (*.edf,*.edif,*.edn,*.edo)
-au BufNewFile,BufRead *.ed\(f\|if\|n\|o\)	setf edif
+" EDIF (*.edf,*.edif,*.edn,*.edo) or edn
+au BufNewFile,BufRead *.ed\(f\|if\|o\)		setf edif
+au BufNewFile,BufRead *.edn
+	\ if getline(1) =~ '^\s*(\s*edif\>' |
+	\   setf edif |
+	\ else |
+	\   setf clojure |
+	\ endif
 
 " EditorConfig (close enough to dosini)
 au BufNewFile,BufRead .editorconfig		setf dosini
```

This sets filetype `clojure` for files with the extension `edn`, unless
the file starts with an EDIF statement (`(edif …)`).

*   edn is specified here: https://github.com/edn-format/edn
*   a description of EDIF is here:
    http://www.rulabinsky.com/cavd/text/chapd.html

Thank you.

**Comment by brammool** [comment]
David Bürgin wrote:

> The following patch adds support for the edn file format.
> 
> ```patch
> diff --git a/runtime/filetype.vim b/runtime/filetype.vim
> index 9c9c808..dba6784 100644
> --- a/runtime/filetype.vim
> +++ b/runtime/filetype.vim
> @@ -676,8 +676,14 @@ au BufNewFile,BufRead *.dtd			setf dtd
>  " DTS/DSTI (device tree files)
>  au BufNewFile,BufRead *.dts,*.dtsi		setf dts
>  
> -" EDIF (*.edf,*.edif,*.edn,*.edo)
> -au BufNewFile,BufRead *.ed\(f\|if\|n\|o\)	setf edif
> +" EDIF (*.edf,*.edif,*.edn,*.edo) or edn
> +au BufNewFile,BufRead *.ed\(f\|if\|o\)		setf edif
> +au BufNewFile,BufRead *.edn
> +	\ if getline(1) =~ '^\s*(\s*edif\>' |
> +	\   setf edif |
> +	\ else |
> +	\   setf clojure |
> +	\ endif
>  
>  " EditorConfig (close enough to dosini)
>  au BufNewFile,BufRead .editorconfig		setf dosini
> ```
> 
> This sets filetype `clojure` for files with the extension `edn`, unless
> the file starts with an EDIF statement (`(edif …)`).
> 
> *   edn is specified here: https://github.com/edn-format/edn
> *   a description of EDIF is here:
>     http://www.rulabinsky.com/cavd/text/chapd.html

Thanks.  There is a chance this doesn't recognize all .edn files that
are edif, I expect anybody who sees that to report it.  Then we can
adjust the pattern.  Perhaps we need to skip empty lines?

-- 
Facepalm statement #1: "I'm going to New York tomorrow, hopefully I have time
to visit the White House"

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by glts** [comment]
> Thanks. There is a chance this doesn't recognize all .edn files that
> are edif, I expect anybody who sees that to report it. Then we can
> adjust the pattern. Perhaps we need to skip empty lines?

I did a GitHub search beforehand, and it looked like EDIF files had the
EDIF statement on the first line (I’m not familiar with EDIF though):

https://github.com/search?o=desc&q=edif+extension%3Aedn&ref=searchresults&s=indexed&type=Code&utf8=%E2%9C%93

**Comment by glts** [comment]
This was fixed in 369b6f5, closing.
